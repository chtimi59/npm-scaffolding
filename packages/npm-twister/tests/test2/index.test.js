const path = require('path')
const childProcess = require('node-child-process-extension')
const fs = require('node-fs-extension')
const nodePrefix = `node ${path.resolve(__dirname, '../../npmt.js')}`
const isUnix = (path.sep === '/')
async function npmt(args="") {
    cmd = `${nodePrefix} ${args}`
    return childProcess.extras.exeToStdOut(cmd, __dirname)
}
const packageJson = path.resolve(__dirname, 'package.json')

const replaceAbsPath = str => {
    const base = __dirname.replace(/\\/g, '\\\\')
    const regEx = new RegExp(`(.*)${base}([^\\s\\"\\']*)([^\\${path.sep}]*)`)
    const arr = str.split('\n').map(i => {
        const u = i.match(regEx)
        if (!u) return i
        const he = u[1] || ""
        const pa = u[2] ? u[2].replace(/\\/g, '/') : ""
        const ta = u[3]  || ""
        return `${he}<ABSOLUTE>${pa}${ta}`
    })
    return arr.join('\n')
}


describe('npm-twister', function () {
    describe('test2', function () {
        beforeAll(async function () {
            fs.extras.writeJsonSync(packageJson, {
                "extends": "a",
                "name": "root-name",
                "description": "root-description",
                "scripts": {
                  "root": "echo Hello"
                },
                "license": "ISC",
                "version": "1.0.0"
            })
        })
        it('empty', async function () {
            try {
                const ret = await npmt("")
                expect(ret).toMatchSnapshot();
            } catch(e) {
                console.error(e)
                throw(e)
            }
        })
        it('local script description', async function () {
            const ret = await npmt("b-hello --help")
            expect(ret).toMatchSnapshot()
        })
        it('run local script A', async function () {
            let ret = replaceAbsPath(await npmt("a 1 2 3"))
            if (isUnix) ret = ret.replace(/1\ 2\ 3/g, "\"1\" \"2\" \"3\"")
            expect(ret).toMatchSnapshot()
        })
        it('run local script B', async function () {
            let ret = replaceAbsPath(await npmt("b 1 2 3"))
            if (isUnix) ret = ret.replace(/1\ 2\ 3/g, "\"1\" \"2\" \"3\"")
            expect(ret).toMatchSnapshot()
        })
        it('run local script C', async function () {
            let ret = replaceAbsPath(await npmt("c 1 2 3"))
            if (isUnix) ret = ret.replace(/1\ 2\ 3/g, "\"1\" \"2\" \"3\"")
            expect(ret).toMatchSnapshot()
        })
        it('run local script a-hello', async function () {
            const ret = await npmt("ha 1 2 3")
            expect(ret).toMatchSnapshot()
        })
        it('bump major version', async function () {
            const ret = await npmt("majversion")
            expect(ret).toMatchSnapshot()
        }, 10000)
        it('run local script b-hello', async function () {
            await npmt("hb")
            const ret = await npmt("a-hello")
            expect(ret).toMatchSnapshot()
        })
    })
})
