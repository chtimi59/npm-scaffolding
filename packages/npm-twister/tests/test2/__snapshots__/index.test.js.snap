// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`npm-twister test2 bump major version 1`] = `
"-- LOAD A ONLY ONCE --
v2.0.0
2.0.0"
`;

exports[`npm-twister test2 empty 1`] = `
"-- LOAD A ONLY ONCE --
Usage:  npmt <command>

wrapped commands from npm (v6.4.1):

    available:
    access, add, add-user, adduser, audit, author, bin, bugs, cache, ci
    completion, config, create, ddp, dedupe, deprecate, dist-tag, dist-tags, docs, doctor
    edit, explore, find, find-dupes, get, help, help-search, home, hook, i
    info, init, innit, install, isntall, issues, it, la, link, list
    ll, ln, login, logout, ls, outdated, owner, pack, ping, prefix
    profile, prune, publish, r, rb, rebuild, remove, repo, restart, rm
    rum, run, run-script, s, se, search, set, show, shrinkwrap, star
    stars, start, stop, t, team, test, token, tst, udpate, un
    uninstall, unlink, unpublish, unstar, up, update, upgrade, urn, version, view
    whoami

    override:
    c, root, v


where:

    b-hello     Some description about b-hello.js
                alias: hb

    b           mapped to package.json scripts

    a-hello     Some description about a-hello.js
                alias: ha

    getversion  get current version
                alias: v

    majversion  update major version

    c           mapped to package.json scripts

    a           mapped to package.json scripts

    root        mapped to package.json scripts

Specific help example:

    npmt install --help

version: 0.0.1"
`;

exports[`npm-twister test2 local script description 1`] = `
"-- LOAD A ONLY ONCE --
Some description about b-hello.js"
`;

exports[`npm-twister test2 run local script A 1`] = `
"-- LOAD A ONLY ONCE --

> root-name@1.0.0 a <ABSOLUTE>
> echo Hello A \\"1\\" \\"2\\" \\"3\\"

Hello A \\"1\\" \\"2\\" \\"3\\""
`;

exports[`npm-twister test2 run local script B 1`] = `
"-- LOAD A ONLY ONCE --

> root-name@1.0.0 b <ABSOLUTE>
> echo Hello B \\"1\\" \\"2\\" \\"3\\"

Hello B \\"1\\" \\"2\\" \\"3\\""
`;

exports[`npm-twister test2 run local script C 1`] = `
"-- LOAD A ONLY ONCE --

> root-name@1.0.0 c <ABSOLUTE>
> echo Hello C \\"1\\" \\"2\\" \\"3\\"

Hello C \\"1\\" \\"2\\" \\"3\\""
`;

exports[`npm-twister test2 run local script a-hello 1`] = `
"-- LOAD A ONLY ONCE --
{ name: 'root-name',
  description: 'root-description',
  scripts:
   { b: 'echo Hello B',
     c: 'echo Hello C',
     a: 'echo Hello A',
     root: 'echo Hello' },
  list:
   [ { un: 1, deux: 1 }, { uno: 1, duos: 2 }, { ein: 1, zwei: 2 } ],
  'from-c': 'Something from c:base',
  license: 'ISC',
  version: '1.0.0' }
[ '1', '2', '3' ]"
`;

exports[`npm-twister test2 run local script b-hello 1`] = `
"-- LOAD A ONLY ONCE --
{ name: 'root-name',
  description: 'root-description',
  scripts:
   { b: 'echo Hello B',
     c: 'echo Hello C',
     a: 'echo Hello A',
     root: 'echo Hello' },
  list: [ { un: 1, deux: 1 }, { uno: 1, duos: 2, tres: 3 }, null ],
  'from-c': 'Something from c:base',
  license: 'ISC',
  version: '2.0.0' }
[]"
`;
